db/view/view_update_generator.cc

```diff
@@ -40,14 +40,21 @@ future<> view_update_generator::start() {
                 _pending_sstables.wait().get();
             }
 
-            // To ensure we don't race with updates, move the entire content
-            // into a local variable.
-            auto sstables_with_tables = std::exchange(_sstables_with_tables, {});
-
-            // If we got here, we will process all tables we know about so far eventually so there
-            // is no starvation
-            for (auto table_it = sstables_with_tables.begin(); table_it != sstables_with_tables.end(); table_it = sstables_with_tables.erase(table_it)) {
-                auto& [t, sstables] = *table_it;
+            // We process a small batch from each table in turn to ensure fair incremental progress.
+            for (auto table_it = _sstables_with_tables.begin(); table_it != _sstables_with_tables.end();) {
+                auto& [t, all_sstables] = *table_it;
+                // Normally, we expect sstables to be added in order, so this
+                // should be a noop. It can help however when scylla is restarted
+                // with staging sstables, as they are not loaded from disk in order.
+                std::sort(all_sstables.begin(), all_sstables.end(), [] (const sstables::shared_sstable& a, const sstables::shared_sstable& b) {
+                    return a->get_first_decorated_key().token() < b->get_first_decorated_key().token();
+                });
```

Comment:
std::ranges::partial_sort can help saving some work, as we need only the first #sstable_batch_size sorted.

Reply:
The work here is not lost, on the next iteration we'll have an almost sorted list.

Reply:
It's not clear that std::sort()'s algorithm has better complexity on almost-sorted lists (some algorithms do, like bubble sort, but I'm sure std::sort() doesn't use bubble sort).

Reply:
Either way, this is a cheap sort: it compares tokens (integers) and moves around shared pointers (cheap to move). Also, even in the bad cases, this list should have elements in the low hundreds.